Support for purchased Volumes on Viz

playmer opened the pull request:
I'm sure there's some additional error handling that needs to be done/moved around, but I think by and large the functionality is there. The approach is likely a bit naïve however, I'm not remotely fluent in JS, so let me know any suggestions to clean it up.

Robonau commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -34,26 +36,87 @@ export default class VizShonenJump extends Connector {
+        var mangaAvailibleByChapters = await this._getMangasAvailibleByChapters();
+
+        mangaAvailibleByVolumes.forEach(mangaAvailibleByVolume => {
+                const index = mangaAvailibleByChapters.findIndex(element => element.title === mangaAvailibleByVolume.title);
```
can just use find and not findindex
```js
            let element = mangaAvailibleByChapters.find(element => element.title === mangaAvailibleByVolume.title);
            // Found it.
            if (element != undefined) {
                var complexManga = JSON.parse(element.id);

                element = {
                    ...
                };
            }
```

playmer replied:
I'll give this a try. I wasn't sure of the semantics allowing me to modify the object inside the array with find, appreciate the tip!

Robonau commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -34,26 +36,87 @@ export default class VizShonenJump extends Connector {
-    async _getChapters(manga) {
+    async _getMangas() {
+        const mangaAvailibleByVolumes = await this._getMangasAvailibleByVolumes();
+        var mangaAvailibleByChapters = await this._getMangasAvailibleByChapters();
```
`let` is better than `var` (this goes for every time you used `var`)

playmer replied:
Thanks!

Robonau commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -10,6 +11,7 @@ export default class VizShonenJump extends Connector {
         this.tags = [ 'manga', 'english' ];
         this.url = 'https://www.viz.com';
         this.requestOptions.headers.set( 'x-referer', this.url );
+        this.EntryTypeEnum = Object.freeze({"volume":1, "chapter":2});
```
is there a reason for the `Object.freeze`, i don't see it, nor can i really test it

playmer replied:
I'm not sure, I just looked up how to do enums in JS and this was recommended. From the docs, it seems like this carries a level of immutability with it, so EntryTypeEnum can't be accidentally modified.: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze

Robonau commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -34,26 +36,87 @@ export default class VizShonenJump extends Connector {
+    }
+
+    async _getMangaChapters(manga) {
+        if ((manga.entry_type & this.EntryTypeEnum.chapter) != this.EntryTypeEnum.chapter) {
```
`if (manga.entry_type  != this.EntryTypeEnum.chapter)`
why use bitwise operators when you don't need to?

actually just use Switch case Statement in _getChapters

actually aren't you trying get from both volumes and chapters?
as the `return chapters.concat(volumes);` in _getChapters would suggest
you currently just get from one or the other

playmer replied:
This is used as a Flag enum, so it's not one or the other, it could be both. A user might have manga chapters available via SJ, as well as volumes they've purchased from Viz. So a switch case doesn't work here, as if both flags were set, the enums value would be `3` and neither condition would be hit.

Thus we union on the values returned from both functions, but sometimes one or the other might be empty.

Robonau replied:
yup this was just be being kinda dumb
`[1,3].includes(this.EntryTypeEnum.chapter)` (or [2,3] )
is probably more readable, using bitwise is rather uncommon in js

or just `this.EntryTypeEnum.chapter != 2` (or 1 depending on vol vs chapters)

Robonau commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -34,26 +36,87 @@ export default class VizShonenJump extends Connector {
+
+    async _getMangaChapters(manga) {
+        if ((manga.entry_type & this.EntryTypeEnum.chapter) != this.EntryTypeEnum.chapter) {
+            console.log(`Exiting from _getMangaChapters ${JSON.stringify(manga)}`);
```
_getMangaVolumes?
don't think you need to stringify either

playmer replied:
This was just for testing purposes, I'm just going to remove it. The stringification was from trying to get something more than like `[Object object]` to print out.

Robonau commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -34,26 +36,86 @@ export default class VizShonenJump extends Connector {
+                            id: complexManga.id,
+                            volume_id: mangaAvailibleByVolume.volume_id,
+                            title: mangaAvailibleByVolume.title,
+                            entry_type: this.EntryTypeEnum.volume | this.EntryTypeEnum.chapter
```
why | and not + ?
(this is what threw me off on the bitwize comment)
`this.EntryTypeEnum.volume + this.EntryTypeEnum.chapter` (or just `3`)
again, bitwise is not common in js

Robonau commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -119,6 +244,7 @@ export default class VizShonenJump extends Connector {
     async _handleConnectorURI(payload) {
         try {
             // API request
+            const testOrigin = new URL(this.url).origin
```
why?

playmer replied:
Superfluous, again for testing, will remove, thanks!

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -101,16 +127,75 @@ export default class VizShonenJump extends Connector {
+        let mangaID = 0;
+
+        if (chapter.id.startsWith("read/manga")) {
+            const pageCountData = await this.requestRegex(responseData, /<div class="mar-b-md"><strong>Length<\/strong> (\d+) pages<\/div>/);
```
The [slightest change with whitespaces](https://regex101.com/r/YflMBl/1/) (linebreak, space) will cause this to fail, isn't there a more reliable way to get the page count?

playmer replied:
Unfortunately not as far as I can tell, I can send an example page to you if you'd like to take a look and see if a better regex could be constructed for this. I'm not much of a wiz at these as you can tell, mostly just built off what I saw already here.

27136773 replied:
Filling in whitespace may be sufficient
`/<div\s+class="mar-b-md">\s*<strong>\s*Length\s*<\/strong>\s+(\d+)\s+pages\s*<\/div>/`
https://regex101.com/r/UjHC6C/1

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -101,16 +127,75 @@ export default class VizShonenJump extends Connector {
+        if (chapter.id.startsWith("read/manga")) {
+            const pageCountData = await this.requestRegex(responseData, /<div class="mar-b-md"><strong>Length<\/strong> (\d+) pages<\/div>/);
+            pageCount = parseInt(pageCountData[0]);
+            const mangaIdData = await this.requestRegex(responseData, /var\s+mangaCommonId\s*=\s*(\d+)/);
```
May be simplified
```javascript
responseData.match(/var\s+mangaCommonId\s*=\s*(\d+)/)[1]
```

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -101,16 +127,75 @@ export default class VizShonenJump extends Connector {
+        }
+    }
+
+    async requestRegex(responseData, regex) {
```
The `async` is obsolete, since there is no `await` anywhere in this method.
Method name contains the phrase `request`, which may be misleading since there is no request performed.
Method seems redundant in general, it can probably be removed in favor of `match`.

playmer replied:
Removing method in favor of `matchAll`, match doesn't seem to return quite what I need.

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -34,7 +34,21 @@ export default class VizShonenJump extends Connector {
-    async _getMangas() {
+    async _getMangasAvailibleByVolumes() {
+        const request = new Request(new URL('/library', this.url), this.requestOptions);
+        let data = await this.fetchDOM(request);
```
Consider using the CSS query directly `this.fetchDOM(request, 'table.purchase-table a')`

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -34,7 +34,21 @@ export default class VizShonenJump extends Connector {
+
+    async _getMangasAvailibleByChapters() {
         const request = new Request(new URL('/shonenjump', this.url), this.requestOptions);
         let data = await this.fetchDOM(request);
```
Consider using the CSS query directly `this.fetchDOM(request, 'div.o_sort_container div.o_sortable a.o_chapters-link')`

playmer replied:
Wouldn't that kill the geolock check?

27136773 replied:
With the current implementation `data` is the body element right now, i think a CSS query is missing to find the chapters `this.fetchDOM(request, 'div.mangas')`

playmer replied:
There is a query to find the chapters: https://github.com/manga-download/hakuneko/blob/d7c1f52adf137762f25a7fed41931fe90731bae3/src/web/mjs/connectors/VizShonenJump.mjs#L55-L72

If I were to implement the fetchDOM with the query, the geolock check wouldn't pass I think right? I'm assuming it'd not contain that text anymore. But I can't test it.

27136773 replied:
Yes you are right, for some reason i didn't see line 59 - 64 during my review ...

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -43,17 +57,29 @@ export default class VizShonenJump extends Connector {
 
+        data = [...data.querySelectorAll('a.o_chapters-link')];
+
         data = [...data.querySelectorAll('div.o_sort_container div.o_sortable a.o_chapters-link')];
```
The method `querySelectorAll` should not be used on an array (`data` is an array at this point, see previous assignment)

playmer replied:
Yeah I think that was a bad rebase on my part, thanks!

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -43,17 +57,29 @@ export default class VizShonenJump extends Connector {
+    async _getMangas() {
+        const mangaAvailibleByVolumes = await this._getMangasAvailibleByVolumes();
+        const mangaAvailibleByChapters = await this._getMangasAvailibleByChapters();
+        let both = mangaAvailibleByVolumes.concat(mangaAvailibleByChapters);
```
May be simplified with `spread` operator.
Also the sort by title is not required if i remember correctly as this is already done during further processing of the manga list.
```javascript
return [ ...mangaAvailibleByVolumes, ...mangaAvailibleByChapters ];
```

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -101,16 +127,75 @@ export default class VizShonenJump extends Connector {
+    async _getMangaVolumes(manga) {
+        let auth = await this._getUserInfo();
+        const request = new Request(new URL(manga.id, this.url), this.requestOptions);
+        let data = await this.fetchDOM(request);
```
Consider using the CSS query directly `this.fetchDOM(request, '#o_products tr td:last-of-type a')`

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -101,16 +127,75 @@ export default class VizShonenJump extends Connector {
+        return [...data.querySelectorAll('#o_products tr td:last-of-type a')]
+            .filter(element => {
+                // login required (e.g. age verification) => javascript:void('login to read');
+                if(/(?=javascript:)(?=.*login)/i.test(element.href)) {
```
Remove unused non-capturing groups from regex `/javascript:.*login/i` should be sufficient

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -101,16 +127,75 @@ export default class VizShonenJump extends Connector {
+                    return auth.isLoggedIn && auth.isAdult;
+                }
+                // subscription required => javascript:void('join to read');
+                if(/(?=javascript:)(?=.*join)/i.test(element.href)) {
```
Remove unused non-capturing groups from regex `/javascript:.*join/i` should be sufficient

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -101,16 +127,75 @@ export default class VizShonenJump extends Connector {
+            .filter(element => {
+                // login required (e.g. age verification) => javascript:void('login to read');
+                if(/(?=javascript:)(?=.*login)/i.test(element.href)) {
+                    return auth.isLoggedIn && auth.isAdult;
```
There is some kind of contradiction, how can there be a login link (`javascript:.*login`), when the user is already logged in (`auth.isLoggedIn === true`)? Is there a use case where this check makes sense?

playmer replied:
I figured this stuff was off. I spent some time and adjusted the volume paths for login handling that's unique to how they work. Had to revert a couple minor changes with the `fetchDom` calls with queries for the checks, but I think this is generally better.

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -101,16 +127,75 @@ export default class VizShonenJump extends Connector {
+            })
+            .map(chapter => {
+                return {
+                    id: chapter.href.replace("hakuneko://cache/", ""),
```
Use `this.getRootRelativeOrAbsoluteURI(chapter, this.url)` instead

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -101,16 +127,75 @@ export default class VizShonenJump extends Connector {
+            .map(chapter => {
+                return {
+                    id: chapter.href.replace("hakuneko://cache/", ""),
+                    title: 'Vol. ' + chapter.href.match(/-volume-([-_0-9]+)/)[1]
```
Is `volume` always lower case or should `i` modifier be added to the regex?

playmer replied:
I think it's always lowercase, but no harm in adding the `i` modifier I think.

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -101,16 +127,75 @@ export default class VizShonenJump extends Connector {
+    async _getChapters(manga) {
+        if (manga.id.startsWith("/shonenjump/chapters")) {
+            return await this._getMangaChapters(manga);
+        } else if (manga.id.startsWith("/library")) {
```
May use `if` instead of `else if` since the method is "short-circuiting" with return.

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -101,16 +127,75 @@ export default class VizShonenJump extends Connector {
+            return await this._getMangaChapters(manga);
+        } else if (manga.id.startsWith("/library")) {
+            return await this._getMangaVolumes(manga);
+        }
```
Missing return (method will await forever), consider to throw if non of the conditions above is fulfilled, e.g.
```javascript
throw new Error(`Failed to get chapters/volumes for manga ${manga.id}, because the manga type is not supported!`);
```

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -101,16 +127,75 @@ export default class VizShonenJump extends Connector {
+            const mangaIdData = await this.requestRegex(responseData, /var\s+mangaCommonId\s*=\s*(\d+)/);
+            mangaID = parseInt(mangaIdData[0]);
+        } else if (chapter.id.startsWith("/shonenjump")) {
+            const data = await this.requestRegex(responseData, /var\s+pages\s*=\s*(\d+)/);
```
Consider using `match`

playmer replied:
Switching to `matchAll`

27136773 reviewed (changes requested):
Approach in general seems good.
Some optional changes possible for code improvement.
Some mandatory changes since they may lead to errors.

playmer committed:
Remove dupe querySelector from bad merge

playmer committed:
Adjust how we're retrieving the url for volumes.

playmer committed:
Switch to `if` for following `else if`since each are returning.

playmer committed:
Throw error in case of a an unsupported manga type (non volume/chapter)

playmer committed:
Remove sort, use spread operator to make the return of getMangas.

playmer committed:
Switch from requestRegex to matchAll

playmer committed:
Remove requestRegex

playmer committed:
Remove unused capture groups

playmer committed:
Call fetchDom with a querySelect directly.

playmer committed:
Another direct call to fetchDom with a query.

playmer committed:
Adds `i` flag to match for volume number.

playmer committed:
Adjusted login handling on volume paths.

playmer committed:
Fix some linting issues.

playmer committed:
Adjust volume page count regex to account for whitespace differences.

27136773 commented on src/web/mjs/connectors/VizShonenJump.mjs:
```diff
@@ -101,16 +123,55 @@ export default class VizShonenJump extends Connector {
+        let data = await this.fetchDOM(request);
+
+        if ( data.innerText.includes('Log in to view your library') ) {
+            alert('Lost your login cookie between fetching the manga list and now (fetching the volume list).', this.label, 'info');
```
Just throw and let HakuNeko do the error handling `throw new Error('Message whats wrong and what a user needs to do ...')`

27136773 commented on src/web/mjs/connectors/anikiga.mjs:
```diff
@@ -1,12 +0,0 @@
-import WordPressMadara from './templates/WordPressMadara.mjs';
-
-export default class Anikiga extends WordPressMadara {
```
Website is still working for me, don't remove

playmer replied:
Oh I thought these didn't stick when I squashed. I'll remove this. 

That being said, I think these two files are a mistake? They're just dupes of two other files with slightly different casing. On a case insensitive file system like Windows you can't have both and so git obviously sees one or the other deleted.

27136773 replied:
Renaming them on your end to fit case on `master` branch may solve the conflict

playmer replied:
I can't rename to match because both files exist in 'master'. So I'm always either removing one or the other. Like I said I can get this uncommitted. But I don't think it's intentional that these dupe files are here.

27136773 commented on src/web/mjs/connectors/mundowuxia.mjs:
```diff
@@ -1,12 +0,0 @@
-import WordPressMadara from './templates/WordPressMadara.mjs';
-
-export default class MundoWuxia extends WordPressMadara {
```
Consider to remove website in a separate pull request (also remove the corresponding icon)

playmer committed:
Initial work, nearly working.

Finish the work of getting pages downloading for volumes

Applying some suggestions from the Draft PR

- Switched `var` to `let`
- Switched to using find instead of findIndex for _getMangas
- Removed extraneous print.
Switch to separate entries for Volumes/Chapters, use the ID to differentiate.

playmer committed:
Change to throwing error so Hakuneko can alert the user.

27136773 reviewed (approved):
Ready to be merged

27136773 reviewed (changes requested):
Check failed, need to fix the `lint` error

playmer committed:
Remove unreachable code

27136773 reviewed (approved):
Looks good now
